docs/code/client_middleware_cookbook.py

```diff
@@ -0,0 +1,132 @@
+            token = self.access_token
+            req.headers["Authorization"] = f"Bearer {token}"
+            resp = await handler(req)
+            if resp.status != 401:
+                return resp
+            async with self.lock:
+                if token != self.access_token:  # Already refreshed
+                    continue
+                url = "https://api.example/refresh"
+                async with req.session.post(url, data=self.refresh_token) as resp:
+                    self.access_token = await resp.json()
+        return resp  # type: ignore[possibly-undefined]
+
+
+class TokenRefreshExpiryMiddleware:
+    def __init__(self, refresh_token: str):
+        self.access_token = ""
+        self.expires_at = 0
+        self.refresh_token = refresh_token
+        self.lock = asyncio.Lock()
+
+    async def __call__(
+        self, req: ClientRequest, handler: ClientHandlerType
+    ) -> ClientResponse:
+        if self.expires_at <= time.time():
+            token = self.access_token
+            async with self.lock:
+                if token == self.access_token:  # Still not refreshed
+                    url = "https://api.example/refresh"
+                    async with req.session.post(url, data=self.refresh_token) as resp:
+                        data = await resp.json()
+                        self.access_token = data["access_token"]
+                        self.expires_at = data["expires_at"]
+
+        req.headers["Authorization"] = f"Bearer {self.access_token}"
+        return await handler(req)
+
+
+async def token_refresh_preemptively_example() -> None:
+    async def set_token(session: ClientSession, event: asyncio.Event) -> None:
+        while True:
+            async with session.post("/refresh") as resp:
+                token = await resp.json()
+                session.headers["Authorization"] = f"Bearer {token['auth']}"
+                event.set()
+                await asyncio.sleep(token["valid_duration"])
+
+    @asynccontextmanager
+    async def auto_refresh_client() -> AsyncIterator[ClientSession]:
+        async with ClientSession() as session:
+            ready = asyncio.Event()
+            t = asyncio.create_task(set_token(session, ready))
+            await ready.wait()
+            yield session
+            t.cancel()
+            with suppress(asyncio.CancelledError):
+                await t
+
+    async with auto_refresh_client() as sess:
+        ...
+
+
+async def ssrf_middleware(
+    req: ClientRequest, handler: ClientHandlerType
+) -> ClientResponse:
+    if req.url.host in {"127.0.0.1", "localhost"}:
+        raise SSRFError(req.url.host)
+    return await handler(req)
+
+
+class SSRFConnector(TCPConnector):
+    async def _resolve_host(
+        self, host: str, port: int, traces: Sequence[Trace] | None = None
+    ) -> list[ResolveResult]:
+        res = await super()._resolve_host(host, port, traces)
+        if any(r["host"] in {"127.0.0.1"} for r in res):
```

Comment:
Maybe add a comment that checking 127.0.0.1 isn't sufficient, and they need to check ::1 etc. Or some warnings that this implementation is incomplete so people don't start using it in production

Reply:
bace3a9a9